12.0 add rental fees

fcayre opened the pull request.

fcayre committed:
First revision with fees modelisation and computation

Computations are run using a dedicated button and change their state
accordingly. Once done, the computation can be reset to draft. This is
required if you want to modify the fees definition, which cannot be
changed if there is at least one non-draft computation.

fcayre committed:
Fix small code style oddities

legalsylvain commented on rental_fees/TODO.rst:
```diff
@@ -0,0 +1,22 @@
+=========
```
si tu veux rester dans du standard OCA, le fichier peut s'appeler ``readme/ROADMAP.rst``
C'est là qu'on met les todo.

fcayre replied:
Merci pour l'info. Je pense que je vais folder tout ça et virer ce fichier de toutes façons, le but de pousser était de partager le code et regarder l'état de la CI après une modif qui impacte d'autres modules, et je suis pas déçu : ça ne passe plus :)

fcayre committed:
t10n fr

fcayre committed:
Add a fees compensation mecanism in case of device breakage or similar

When such an event occurs, the fees stop and are replaced by a full device
compensation payout.
The amount of the payout is the standard price of the device, computed here
using the purchase price of the device and the reduction ratio agreed in the
fees agreement.

fcayre committed:
Add computed fields and UI sugar to ease user search/ filtering/ grouping

fcayre committed:
fixup: split views file

fcayre committed:
fixup: rename computation model file

fcayre committed:
Add py3o odt report for rental fees

Also fix computation invoice amount, which was also taking future
invoices into account instead of just past computations invoices.

fcayre committed:
t10n fr update

fcayre committed:
fixup missing report_py3o module dependency

fcayre committed:
Fix and refactor fees in case of device scrapping

Compensation is now detected based on stock moves, not using
the dedicated project: this way, partial device breakage is
not mode a false positive scrap.

The amount of the compensation is higher than the standard
device price when fees + purchase price are, which is a fix
compared to previous implementation.

fcayre committed:
Small refactoring to lower the number of db requests

fcayre committed:
Remove useless method argument

fees definition is an attribute of the fees computation: no need to pass it to the split_periods_wrt_fees_def method.

fcayre committed:
Refactor computation add details method

They do not need to return the created object, as the total fees computation can be performed
at the very end from the database content.

fcayre committed:
Fix a bug when a fees def line has no end date (infinite)

fcayre committed:
Add a no rental compensation mecanism

fcayre committed:
Deny fees def updates of non crucial fields only when a computation is present

fcayre committed:
fixup compensation doc

fcayre committed:
fr t10n update

fcayre committed:
Fix ##DATE## marker lang in fees invoice lines

fcayre committed:
Fix fees computation invoice amount computation to allow refund invoices

fcayre committed:
Review fees computation tree and form views

fcayre committed:
Improve fees definition view: display ratio field as percentage

fcayre committed:
Set fees computation fields as readonly once the no more draft

fcayre committed:
Report: add table to highlight alltime fees while clearly showing fees to invoice

fcayre committed:
Fix po line from device request in fees_def prices method

fcayre committed:
Report: replace useless contract column by compensation reason

fcayre committed:
Report: add a per-device revenue section and improve overall styling

fcayre committed:
Fix compensation fees

Do not forget we already paid the purchase price, so the compensation
price should be the max of the already paid fees and the difference
between standard price and purchase price.

fcayre committed:
Report: Fix first table column alignments

fcayre committed:
First version of a fees spreadsheet report

fcayre commented:
/oca merge patch

fcayre commented:
/ocabot merge patch
